aptos-move/aptos-vm-logging/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod counters;
+pub mod log_schema;
+
+pub mod prelude {
+    pub use crate::{
+        counters::CRITICAL_ERRORS, speculative_debug, speculative_error, speculative_info,
+        speculative_log, speculative_trace, speculative_warn,
+    };
+}
+
+use crate::{counters::CRITICAL_ERRORS, log_schema::AdapterLogSchema};
+use aptos_logger::{prelude::*, Level};
+use aptos_speculative_state_helper::{SpeculativeEvent, SpeculativeEvents};
+use arc_swap::ArcSwapOption;
+use once_cell::sync::Lazy;
+use std::sync::Arc;
+
+struct VMLogEntry {
+    level: Level,
+    context: AdapterLogSchema,
+    message: String,
+}
+
+impl VMLogEntry {
+    fn new(level: Level, context: AdapterLogSchema, message: String) -> Self {
+        Self {
+            level,
+            context,
+            message,
+        }
+    }
+}
+
+// Implement dispatching for a recorded VM log entry to support recording
+// as speculative logging event (and eventual dispatching as needed).
+impl SpeculativeEvent for VMLogEntry {
+    fn dispatch(self) {
+        match self.level {
+            Level::Error => {
+                // TODO: Consider using SpeculativeCounter to increase CRITICAL_ERRORS
+                // on the critical path instead of async dispatching.
+                error!(self.context, "{}", self.message);
+                CRITICAL_ERRORS.inc();
```

Comment:
hmm, is every single error message in vm considered critical error?

Reply:
It was like this so I kept the behavior, but yeah, weird. Why not just alert on "error"